### PR TITLE
chore: Reduce memory consumption of AMVF

### PR DESCRIPTION
**Pull request by paulgessinger**
This pushes a number of vertex calculations to separate `.cpp` files so they can be compiled once and linked. This switches a couple of dimension checks to runtime checks. I didn't do a runtime test yet, but that's something I need to check.

The memory consumption goes down by around 30%, while adding two new compilation units of about 900M each.

### CODE REVIEW

**Comment by paulgessinger** [comment]
### Feature
| file  |                                                                      max_rss [M] |   time [s] |
| ------------------------------------------------------------------------ | ------------- | ---------- |
| Tests/UnitTests/Core/Vertexing/IterativeVertexFinderTests.cpp           |        1344.44 |      15.58 | 
| Examples/Algorithms/Vertexing/src/IterativeVertexFinderAlgorithm.cpp    |        1280.00 |      12.22 | 
| Tests/UnitTests/Core/Vertexing/AdaptiveMultiVertexFinderTests.cpp       |        1263.67 |      17.66 | 
| Examples/Algorithms/Vertexing/src/AdaptiveMultiVertexFinderAlgorithm.cpp|        1258.39 |      12.58 | 
| Examples/Algorithms/Vertexing/src/VertexFitterAlgorithm.cpp             |        1162.40 |      10.50 | 
| Tests/UnitTests/Core/Vertexing/FullBilloirVertexFitterTests.cpp         |        1109.15 |      11.16 | 
| Tests/UnitTests/Core/Vertexing/AdaptiveMultiVertexFitterTests.cpp       |        1047.30 |      10.96 | 
| Core/src/Vertexing/KalmanVertexUpdater.cpp                              |        1013.28 |       7.99 | 
| Core/src/Vertexing/KalmanVertexTrackUpdater.cpp                         |         975.01 |       8.22 | 
| Tests/UnitTests/Core/Vertexing/ZScanVertexFinderTests.cpp               |         948.68 |       8.34 | 

|file |                                                                              max_rss [M] |   time [s]|
|------------------------------------------------------------------------------- | -------------  | ----------| 
|Tests/UnitTests/Core/Vertexing/AdaptiveMultiVertexFinderTests.cpp                 |     1263.67   |    17.66 |
|Tests/UnitTests/Core/Vertexing/IterativeVertexFinderTests.cpp                     |     1344.44   |    15.58 |
|Examples/Algorithms/Vertexing/src/AdaptiveMultiVertexFinderAlgorithm.cpp          |     1258.39   |    12.58 |
|Examples/Algorithms/Vertexing/src/IterativeVertexFinderAlgorithm.cpp              |     1280.00   |    12.22 |
|Tests/UnitTests/Core/Vertexing/FullBilloirVertexFitterTests.cpp                   |     1109.15   |    11.16 |
|Tests/UnitTests/Core/Vertexing/AdaptiveMultiVertexFitterTests.cpp                 |     1047.30   |    10.96 |
|Examples/Algorithms/Vertexing/src/VertexFitterAlgorithm.cpp                       |     1162.40   |    10.50 |
|Examples/Io/Performance/ActsExamples/Io/Performance/VertexPerformanceWriter.cpp   |      910.57   |     9.57 |
|Tests/UnitTests/Core/Vertexing/ImpactPointEstimatorTests.cpp                      |      945.42   |     9.00 |
|Examples/Python/src/Vertexing.cpp                                                 |      947.34   |     8.60 |

### Main 

| file                                                                        | max_rss [M]  |   time [s] |
| ------------------------------------------------------------------------ | ------------- | ---------- |
| Examples/Algorithms/Vertexing/src/AdaptiveMultiVertexFinderAlgorithm.cpp   |     1760.58  |     20.28 |
| Tests/UnitTests/Core/Vertexing/AdaptiveMultiVertexFinderTests.cpp          |     1675.89  |     28.81 |
| Tests/UnitTests/Core/Vertexing/AdaptiveMultiVertexFitterTests.cpp          |     1525.40  |     19.84 |
| Tests/UnitTests/Core/Vertexing/IterativeVertexFinderTests.cpp              |     1334.74  |     15.69 |
| Examples/Algorithms/Vertexing/src/IterativeVertexFinderAlgorithm.cpp       |     1249.56  |     12.15 |
| Tests/UnitTests/Core/Vertexing/KalmanVertexTrackUpdaterTests.cpp           |     1204.06  |     10.18 |
| Examples/Algorithms/Vertexing/src/VertexFitterAlgorithm.cpp                |     1117.44  |     10.58 |
| Tests/UnitTests/Core/Vertexing/KalmanVertexUpdaterTests.cpp                |     1091.70  |      9.11 |
| Tests/UnitTests/Core/Vertexing/FullBilloirVertexFitterTests.cpp            |     1038.19  |     10.87 |
| Tests/UnitTests/Core/Vertexing/ZScanVertexFinderTests.cpp                  |      889.13  |      9.00 |

| file                                    | max_rss [M] | time [s] |
| ----------------------------------------| -------------| --------------------------------------------------- |
| Tests/UnitTests/Core/Vertexing/AdaptiveMultiVertexFinderTests.cpp                   |   1675.89 |      28.81 |
| Examples/Algorithms/Vertexing/src/AdaptiveMultiVertexFinderAlgorithm.cpp            |   1760.58 |      20.28 |
| Tests/UnitTests/Core/Vertexing/AdaptiveMultiVertexFitterTests.cpp                   |   1525.40 |      19.84 |
| Tests/UnitTests/Core/Vertexing/IterativeVertexFinderTests.cpp                       |   1334.74 |      15.69 |
| Examples/Algorithms/Vertexing/src/IterativeVertexFinderAlgorithm.cpp                |   1249.56 |      12.15 |
| Tests/UnitTests/Core/Vertexing/FullBilloirVertexFitterTests.cpp                     |   1038.19 |      10.87 |
| Examples/Algorithms/Vertexing/src/VertexFitterAlgorithm.cpp                         |   1117.44 |      10.58 |
| Tests/UnitTests/Core/Vertexing/KalmanVertexTrackUpdaterTests.cpp                    |   1204.06 |      10.18 |
| Examples/Io/Performance/ActsExamples/Io/Performance/VertexPerformanceWriter.cpp     |    841.63 |       9.26 |
| Tests/UnitTests/Core/Vertexing/KalmanVertexUpdaterTests.cpp                         |   1091.70 |       9.11 |

File: Core/include/Acts/Vertexing/TrackAtVertex.hpp [hunk 1]
```diff
@@ -74,4 +74,24 @@ struct TrackAtVertex {
   bool isLinearized = false;
 };
 
+struct TrackAtVertexRef {
```
Comment (paulgessinger) [inline Core/include/Acts/Vertexing/TrackAtVertex.hpp:77]: Now this is a bit of a stopgap solution to replacing track at vertex completely with a type erasure type. This autoconstructs from a track at vertex, and maintains references.

**Comment by paulgessinger** [comment]
Runtime of the ckf tracking physmon workflow seems unaffected (100 events, 30 runs each, no IO).

## Main

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `python CI/physmon/workflows/physmon_ckf_tracking.py /tmp/out` | 21.353 ± 0.591 | 20.423 | 22.697 | 1.00 |

## This PR
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `python CI/physmon/workflows/physmon_ckf_tracking.py /tmp/out` | 21.169 ± 0.532 | 20.432 | 22.712 | 1.00 |

**Review by andiwand — APPROVED** [review]

File: Core/include/Acts/Vertexing/TrackAtVertex.hpp (hunk 1 above)
Comment (felix-russo) [reply Core/include/Acts/Vertexing/TrackAtVertex.hpp:77]: I don't quite understand why this is needed, could you explain?

**Comment by acts-project-service** [comment]
## :white_check_mark: Athena integration test results [69290bed30a2f3dcc3c144b9949f918a4e4fb14e]
### :white_check_mark: All tests successful

|status|job|report|
|------|---|------|
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34770497)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsBenchmarkWithSpot.sh 8 100](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34770496)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsWorkflow.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34770495)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateAmbiguityResolution.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34770494)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateResolvedTracks.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34770493)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateTracks.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34770492)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateActsCoreSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34770491)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateActsSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34770490)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34770489)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateOrthogonalSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34770488)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateClusters.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34770487)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsPersistifyEDM.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34770486)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsGSFRefitting.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34770485)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsKfRefitting.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34770484)||
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsExtrapolationAlgTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34770483)||
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsITkTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34770482)||
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34770481)||
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34770480)||
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34770479)||
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34770478)||
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34770477)||
|:green_circle: | [run_art_test: test_data18_13TeV_1000evt](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34770476)||
|:green_circle: | [run_art_test: test_ttbarPU40_reco](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34770475)||

**Pull request merged**